File: python/receipt_processor.py
#!/usr/bin/env python3
"""
Receipt Processor — Local Python alternative.

Watches a folder for receipt images/PDFs, extracts text via OCR,
and logs expenses to an Excel spreadsheet.

Usage:
    python receipt_processor.py                    # Process ~/Receipts once
    python receipt_processor.py --watch            # Watch folder continuously
    python receipt_processor.py --folder /path     # Custom folder
    python receipt_processor.py --spreadsheet out.xlsx
"""

import argparse
import os
import re
import shutil
import sys
import time
from datetime import datetime
from pathlib import Path

try:
    import pytesseract
    from PIL import Image
except ImportError:
    print("Install dependencies: pip install -r requirements.txt")
    sys.exit(1)

try:
    from openpyxl import Workbook, load_workbook
    from openpyxl.styles import Font, Alignment
except ImportError:
    print("Install openpyxl: pip install openpyxl")
    sys.exit(1)

try:
    import fitz  # PyMuPDF
    HAS_PDF = True
except ImportError:
    HAS_PDF = False

HEADERS = ["Date", "Vendor", "Amount", "Category", "Payment Method",
           "File Name", "File Path", "Notes"]

CATEGORY_PATTERNS = [
    (r"grocery|kroger|walmart|costco|safeway|trader joe|whole foods|aldi", "Groceries"),
    (r"restaurant|cafe|coffee|starbucks|mcdonald|burger|pizza|diner", "Dining"),
    (r"gas|shell|chevron|exxon|mobil|fuel|\bbp\b", "Gas"),
    (r"pharmacy|cvs|walgreens|rx|prescription|medical|doctor|clinic", "Healthcare"),
    (r"amazon|target|best buy|home depot|lowes|ikea", "Shopping"),
    (r"electric|water|internet|phone|utility|at&t|verizon|comcast", "Utilities"),
    (r"hotel|airbnb|airline|flight|uber|lyft", "Travel"),
]

PAYMENT_PATTERNS = [
    (r"visa", "Visa"),
    (r"mastercard|mc", "Mastercard"),
    (r"amex|american express", "Amex"),
    (r"discover", "Discover"),
    (r"debit", "Debit"),
    (r"cash", "Cash"),
    (r"apple pay", "Apple Pay"),
]


def extract_text_from_image(path: Path) -> str:
    img = Image.open(path)
    return pytesseract.image_to_string(img)


def extract_text_from_pdf(path: Path) -> str:
    if not HAS_PDF:
        print(f"  Skipping PDF {path.name} (install PyMuPDF: pip install pymupdf)")
        return ""
    doc = fitz.open(path)
    text = ""
    for page in doc:
        text += page.get_text()
    doc.close()
    return text


def extract_text(path: Path) -> str:
    suffix = path.suffix.lower()
    if suffix in (".jpg", ".jpeg", ".png", ".heic", ".tiff", ".bmp", ".webp"):
        return extract_text_from_image(path)
    elif suffix == ".pdf":
        return extract_text_from_pdf(path)
    return ""


def parse_receipt(text: str) -> dict:
    result = {"vendor": "", "amount": "", "category": "", "payment_method": ""}

    lines = [l.strip() for l in text.split("\n") if l.strip()]
    if lines:
        result["vendor"] = lines[0][:50]

    total_patterns = [
        r"(?:total|grand total|amount due|balance due|total due)\s*[:\s]*\$?\s*(\d+[.,]\d{2})",
        r"\$\s*(\d+[.,]\d{2})\s*$",
        r"(\d+[.,]\d{2})\s*(?:total|due)",
    ]
    for pattern in total_patterns:
        match = re.search(pattern, text, re.IGNORECASE | re.MULTILINE)
        if match:
            result["amount"] = match.group(1).replace(",", ".")
            break

    for pattern, method in PAYMENT_PATTERNS:
        if re.search(pattern, text, re.IGNORECASE):
            result["payment_method"] = method
            break

    for pattern, category in CATEGORY_PATTERNS:
        if re.search(pattern, text, re.IGNORECASE):
            result["category"] = category
            break

    return result


def get_month_folder(base: Path) -> Path:
    now = datetime.now()
    month_names = [
        "January", "February", "March", "April", "May", "June",
        "July", "August", "September", "October", "November", "December",
    ]
    year_folder = base / str(now.year)
    month_folder = year_folder / f"{now.month:02d}-{month_names[now.month - 1]}"
    month_folder.mkdir(parents=True, exist_ok=True)
    return month_folder


def rename_receipt(path: Path, parsed: dict) -> str:
    date_str = datetime.now().strftime("%Y-%m-%d")
    vendor = re.sub(r'[^\w\s-]', '', parsed["vendor"]).strip().replace(" ", "_")[:30]
    amount = parsed["amount"] if parsed["amount"] else "unknown"
    suffix = path.suffix
    return f"{date_str}_{vendor}_${amount}{suffix}"


def init_spreadsheet(path: Path) -> Workbook:
    if path.exists():
        return load_workbook(path)

    wb = Workbook()
    ws = wb.active
    ws.title = "Expenses"
    for col, header in enumerate(HEADERS, 1):
        cell = ws.cell(row=1, column=col, value=header)
        cell.font = Font(bold=True)
        cell.alignment = Alignment(horizontal="center")
    ws.column_dimensions["A"].width = 12
    ws.column_dimensions["B"].width = 25
    ws.column_dimensions["C"].width = 12
    ws.column_dimensions["D"].width = 15
    ws.column_dimensions["E"].width = 18
    ws.column_dimensions["F"].width = 40
    ws.column_dimensions["G"].width = 50
    wb.save(path)
    return wb


def log_to_spreadsheet(spreadsheet_path: Path, data: dict, file_name: str, file_path: str):
    wb = init_spreadsheet(spreadsheet_path)
    ws = wb["Expenses"]
    row = [
        datetime.now().strftime("%Y-%m-%d"),
        data["vendor"],
        float(data["amount"]) if data["amount"] else "",
        data["category"],
        data["payment_method"],
        file_name,
        file_path,
        "",
    ]
    ws.append(row)
    wb.save(spreadsheet_path)


def process_file(file_path: Path, base_folder: Path, spreadsheet_path: Path) -> bool:
    print(f"  Processing: {file_path.name}")
    text = extract_text(file_path)
    if not text.strip():
        print(f"    No text extracted, skipping")
        return False

    parsed = parse_receipt(text)
    new_name = rename_receipt(file_path, parsed)
    month_folder = get_month_folder(base_folder)
    dest = month_folder / new_name

    # Handle duplicates
    counter = 1
    while dest.exists():
        stem = dest.stem + f"_{counter}"
        dest = month_folder / (stem + dest.suffix)
        counter += 1

    shutil.move(str(file_path), str(dest))
    log_to_spreadsheet(spreadsheet_path, parsed, new_name, str(dest))

    print(f"    Vendor: {parsed['vendor']}")
    print(f"    Amount: ${parsed['amount']}")
    print(f"    Category: {parsed['category']}")
    print(f"    Saved to: {dest}")
    return True


def process_folder(folder: Path, spreadsheet_path: Path):
    extensions = {".jpg", ".jpeg", ".png", ".heic", ".tiff", ".bmp", ".webp", ".pdf"}
    files = [f for f in folder.iterdir()
             if f.is_file() and f.suffix.lower() in extensions]

    if not files:
        print("No receipt files found in inbox.")
        return

    print(f"Found {len(files)} receipt(s) to process.\n")
    processed = 0
    for f in sorted(files):
        if process_file(f, folder, spreadsheet_path):
            processed += 1

    print(f"\nDone. Processed {processed}/{len(files)} receipts.")
    print(f"Spreadsheet: {spreadsheet_path}")


def watch_folder(folder: Path, spreadsheet_path: Path, interval: int = 30):
    print(f"Watching {folder} for new receipts (every {interval}s)...")
    print("Press Ctrl+C to stop.\n")
    seen = set()
    extensions = {".jpg", ".jpeg", ".png", ".heic", ".tiff", ".bmp", ".webp", ".pdf"}

    while True:
        files = [f for f in folder.iterdir()
                 if f.is_file() and f.suffix.lower() in extensions and f.name not in seen]
        for f in files:
            seen.add(f.name)
            process_file(f, folder, spreadsheet_path)
        time.sleep(interval)


def main():
    parser = argparse.ArgumentParser(description="Process receipt images into organized folders + Excel spreadsheet")
    parser.add_argument("--folder", type=str, default=str(Path.home() / "Receipts"),
                        help="Receipt folder path (default: ~/Receipts)")
    parser.add_argument("--spreadsheet", type=str, default=None,
                        help="Excel file path (default: <folder>/expense_log.xlsx)")
    parser.add_argument("--watch", action="store_true",
                        help="Continuously watch folder for new receipts")
    parser.add_argument("--interval", type=int, default=30,
                        help="Watch interval in seconds (default: 30)")
    args = parser.parse_args()

    folder = Path(args.folder)
    folder.mkdir(parents=True, exist_ok=True)

    spreadsheet = Path(args.spreadsheet) if args.spreadsheet else folder / "expense_log.xlsx"

    print(f"Receipt folder: {folder}")
    print(f"Spreadsheet: {spreadsheet}\n")

    if args.watch:
        watch_folder(folder, spreadsheet, args.interval)
    else:
        process_folder(folder, spreadsheet)


if __name__ == "__main__":
    main()

File: google-apps-script/ReceiptProcessor.js
/**
 * Google Apps Script: Receipt Processor
 *
 * Watches a Google Drive folder for new receipt images/PDFs,
 * extracts text via OCR, and logs expenses to a Google Sheet.
 *
 * SETUP:
 * 1. Create a Google Sheet named "Expense Tracker"
 * 2. Create a Google Drive folder named "Receipts"
 * 3. Open Extensions > Apps Script in the sheet
 * 4. Paste this code and save
 * 5. Run setupTrigger() once to enable automatic processing
 * 6. Run setupSheet() once to create headers
 */

const CONFIG = {
  RECEIPT_FOLDER_NAME: 'Receipts',
  SPREADSHEET_NAME: 'Expense Tracker',
  SHEET_NAME: 'Expenses',
  PROCESSED_PROPERTY: 'processedFiles',
};

function setupSheet() {
  const ss = SpreadsheetApp.getActiveSpreadsheet();
  let sheet = ss.getSheetByName(CONFIG.SHEET_NAME);
  if (!sheet) {
    sheet = ss.insertSheet(CONFIG.SHEET_NAME);
  }

  const headers = [
    'Date Added',
    'File Name',
    'Vendor (extracted)',
    'Amount (extracted)',
    'Category',
    'Payment Method',
    'OCR Text',
    'File Link',
    'Notes',
  ];

  sheet.getRange(1, 1, 1, headers.length).setValues([headers]);
  sheet.getRange(1, 1, 1, headers.length).setFontWeight('bold');
  sheet.setFrozenRows(1);

  sheet.setColumnWidth(1, 120); // Date
  sheet.setColumnWidth(2, 200); // File Name
  sheet.setColumnWidth(3, 150); // Vendor
  sheet.setColumnWidth(4, 100); // Amount
  sheet.setColumnWidth(5, 120); // Category
  sheet.setColumnWidth(6, 130); // Payment
  sheet.setColumnWidth(7, 300); // OCR Text
  sheet.setColumnWidth(8, 200); // Link
}

function setupTrigger() {
  // Remove existing triggers to avoid duplicates
  ScriptApp.getProjectTriggers().forEach(trigger => {
    if (trigger.getHandlerFunction() === 'processNewReceipts') {
      ScriptApp.deleteTrigger(trigger);
    }
  });

  // Check for new receipts every 15 minutes
  ScriptApp.newTrigger('processNewReceipts')
    .timeDriven()
    .everyMinutes(15)
    .create();

  Logger.log('Trigger created: processNewReceipts runs every 15 minutes');
}

function processNewReceipts() {
  const folder = getOrCreateFolder(CONFIG.RECEIPT_FOLDER_NAME);
  const processed = getProcessedFiles();
  const files = folder.getFiles();

  while (files.hasNext()) {
    const file = files.next();
    const fileId = file.getId();

    if (processed.has(fileId)) continue;

    try {
      processReceiptFile(file);
      processed.add(fileId);
    } catch (e) {
      Logger.log('Error processing ' + file.getName() + ': ' + e.message);
    }
  }

  // Also check monthly subfolders
  const subfolders = folder.getFolders();
  while (subfolders.hasNext()) {
    const subfolder = subfolders.next();
    const subfiles = subfolder.getFiles();
    while (subfiles.hasNext()) {
      const file = subfiles.next();
      const fileId = file.getId();
      if (processed.has(fileId)) continue;
      try {
        processReceiptFile(file);
        processed.add(fileId);
      } catch (e) {
        Logger.log('Error processing ' + file.getName() + ': ' + e.message);
      }
    }
  }

  saveProcessedFiles(processed);
}

function processReceiptFile(file) {
  const mimeType = file.getMimeType();
  let ocrText = '';

  if (mimeType.startsWith('image/')) {
    ocrText = extractTextFromImage(file);
  } else if (mimeType === 'application/pdf') {
    ocrText = extractTextFromPdf(file);
  } else {
    Logger.log('Skipping unsupported file type: ' + mimeType);
    return;
  }

  const extracted = parseReceiptText(ocrText);
  logToSheet(file, extracted, ocrText);

  // Move to monthly subfolder
  organizeIntoMonthFolder(file);
}

function extractTextFromImage(file) {
  // Use Google Drive's built-in OCR by converting image to Google Doc
  const resource = {
    title: 'OCR_TEMP_' + file.getName(),
    mimeType: file.getMimeType(),
  };

  const options = {
    ocr: true,
    ocrLanguage: 'en',
  };

  const blob = file.getBlob();
  const ocrFile = Drive.Files.insert(resource, blob, options);
  const doc = DocumentApp.openById(ocrFile.id);
  const text = doc.getBody().getText();

  // Clean up temp file
  DriveApp.getFileById(ocrFile.id).setTrashed(true);

  return text;
}

function extractTextFromPdf(file) {
  const resource = {
    title: 'OCR_TEMP_' + file.getName(),
    mimeType: 'application/pdf',
  };

  const options = {
    ocr: true,
    ocrLanguage: 'en',
  };

  const blob = file.getBlob();
  const ocrFile = Drive.Files.insert(resource, blob, options);
  const doc = DocumentApp.openById(ocrFile.id);
  const text = doc.getBody().getText();

  DriveApp.getFileById(ocrFile.id).setTrashed(true);

  return text;
}

function parseReceiptText(text) {
  const result = {
    vendor: '',
    amount: '',
    category: '',
    paymentMethod: '',
  };

  // Extract vendor: usually the first non-empty line
  const lines = text.split('\n').filter(line => line.trim().length > 0);
  if (lines.length > 0) {
    result.vendor = lines[0].trim().substring(0, 50);
  }

  // Extract total amount: look for patterns like "Total: $XX.XX" or "TOTAL $XX.XX"
  const totalPatterns = [
    /(?:total|grand total|amount due|balance due|total due)\s*[:\s]*\$?\s*(\d+[.,]\d{2})/i,
    /\$\s*(\d+[.,]\d{2})\s*$/m,
    /(\d+[.,]\d{2})\s*(?:total|due)/i,
  ];

  for (const pattern of totalPatterns) {
    const match = text.match(pattern);
    if (match) {
      result.amount = match[1].replace(',', '.');
      break;
    }
  }

  // Detect payment method
  const paymentPatterns = [
    { pattern: /visa/i, method: 'Visa' },
    { pattern: /mastercard|mc/i, method: 'Mastercard' },
    { pattern: /amex|american express/i, method: 'Amex' },
    { pattern: /discover/i, method: 'Discover' },
    { pattern: /debit/i, method: 'Debit' },
    { pattern: /cash/i, method: 'Cash' },
    { pattern: /apple pay/i, method: 'Apple Pay' },
  ];

  for (const { pattern, method } of paymentPatterns) {
    if (pattern.test(text)) {
      result.paymentMethod = method;
      break;
    }
  }

  // Guess category from common keywords
  const categoryPatterns = [
    { pattern: /grocery|kroger|walmart|costco|safeway|trader joe|whole foods|aldi/i, category: 'Groceries' },
    { pattern: /restaurant|cafe|coffee|starbucks|mcdonald|burger|pizza|diner/i, category: 'Dining' },
    { pattern: /gas|shell|chevron|exxon|mobil|fuel|bp\b/i, category: 'Gas' },
    { pattern: /pharmacy|cvs|walgreens|rx|prescription|medical|doctor|clinic/i, category: 'Healthcare' },
    { pattern: /amazon|target|best buy|home depot|lowes|ikea/i, category: 'Shopping' },
    { pattern: /electric|water|internet|phone|utility|at&t|verizon|comcast/i, category: 'Utilities' },
    { pattern: /hotel|airbnb|airline|flight|uber|lyft/i, category: 'Travel' },
  ];

  for (const { pattern, category } of categoryPatterns) {
    if (pattern.test(text)) {
      result.category = category;
      break;
    }
  }

  return result;
}

function logToSheet(file, extracted, ocrText) {
  const ss = SpreadsheetApp.getActiveSpreadsheet();
  let sheet = ss.getSheetByName(CONFIG.SHEET_NAME);
  if (!sheet) {
    setupSheet();
    sheet = ss.getSheetByName(CONFIG.SHEET_NAME);
  }

  const row = [
    new Date(),
    file.getName(),
    extracted.vendor,
    extracted.amount ? parseFloat(extracted.amount) : '',
    extracted.category,
    extracted.paymentMethod,
    ocrText.substring(0, 500),
    file.getUrl(),
    '',
  ];

  sheet.appendRow(row);
}

function organizeIntoMonthFolder(file) {
  const parentFolder = getOrCreateFolder(CONFIG.RECEIPT_FOLDER_NAME);
  const now = new Date();
  const year = now.getFullYear().toString();
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const monthNames = [
    'January', 'February', 'March', 'April', 'May', 'June',
    'July', 'August', 'September', 'October', 'November', 'December',
  ];
  const monthFolderName = month + '-' + monthNames[now.getMonth()];

  let yearFolder = getOrCreateSubfolder(parentFolder, year);
  let monthFolder = getOrCreateSubfolder(yearFolder, monthFolderName);

  // Move file if not already in the month folder
  const parents = file.getParents();
  while (parents.hasNext()) {
    const parent = parents.next();
    if (parent.getId() !== monthFolder.getId()) {
      monthFolder.addFile(file);
      parent.removeFile(file);
    }
  }
}

function getOrCreateFolder(name) {
  const folders = DriveApp.getFoldersByName(name);
  if (folders.hasNext()) {
    return folders.next();
  }
  return DriveApp.createFolder(name);
}

function getOrCreateSubfolder(parent, name) {
  const folders = parent.getFoldersByName(name);
  if (folders.hasNext()) {
    return folders.next();
  }
  return parent.createFolder(name);
}

function getProcessedFiles() {
  const props = PropertiesService.getScriptProperties();
  const data = props.getProperty(CONFIG.PROCESSED_PROPERTY);
  return new Set(data ? JSON.parse(data) : []);
}

function saveProcessedFiles(processed) {
  const props = PropertiesService.getScriptProperties();
  props.setProperty(CONFIG.PROCESSED_PROPERTY, JSON.stringify([...processed]));
}

/**
 * Manual trigger: process all unprocessed receipts now.
 * Run this from the Apps Script editor to process immediately.
 */
function processNow() {
  processNewReceipts();
  Logger.log('Manual processing complete.');
}
